File: scala/bob/src/main/scala/bob.scala
class Bob {

  def hey(phrase: String): String = phrase.trim() match {
    case phrase if silence(phrase) => "Fine. Be that way!"
    case phrase if shouting(phrase) => "Whoa, chill out!"
    case phrase if questioning(phrase) => "Sure."
    case _ => "Whatever."
  }

  def silence(phrase: String): Boolean = {
    phrase == ""
  }

  def shouting(phrase: String): Boolean = {
    phrase.exists(_.isLetter) && !phrase.exists(Character.isLowerCase(_))
  }

  def questioning(phrase: String): Boolean = {
    phrase.endsWith("?")
  }
}

File: java/bob/src/main/java/Bob.java
class Bob {

  public String hey(String phrase) {
    String response = "Whatever.";

    if (isSilence(phrase)) {
      response = "Fine. Be that way!";
    } else if (isShouting(phrase)) {
      response = "Whoa, chill out!";
    } else if (isQuestion(phrase)) {
      response = "Sure.";
    }

    return response;
  }

  private boolean isSilence(String phrase) {
    return phrase.trim().equals("");
  }

  private boolean isQuestion(String phrase) {
    return phrase.endsWith("?");
  }

  private boolean isShouting(String phrase) {
    return !isNumbersOnly(phrase) && phrase.toUpperCase().equals(phrase);
  }

  private boolean isNumbersOnly(String phrase) {
    return phrase.chars().allMatch(c -> Character.isDigit(c) || Character.isSpaceChar(c) || c == ',' || c == '?');
  }

}
